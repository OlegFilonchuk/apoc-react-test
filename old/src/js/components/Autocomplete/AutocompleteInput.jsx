import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Input from '../Input/Input';

import './AutocompleteInput.less';
import { getDataTestElementProps, dataTestElementPropTypes } from '../../utils/dataTestElementPropUtils';

const BASE_CLASS = 'autocomplete-input';

export default class AutocompleteInput extends Component {
  onInputClick = e => {
    this.props.onClick(e, this.expandedValue);
    this.props.showFilteredOptions(this.props.value);
  };

  onButtonClick = e => {
    if (!this.props.disabled) {
      this.props.onClick(e, this.expandedValue);
      this.props.showAllOptions();
    }
  };

  onInputRef = ref => {
    this.root = ref;
  };

  get expandedValue() {
    return !(this.root !== document.activeElement && this.props.isExpanded);
  }

  get errors() {
    return this.props.children ? <div className="errors">{this.props.children}</div> : null;
  }

  get wrapperClass() {
    return classNames({
      disabled: this.props.disabled,
      [BASE_CLASS]: true,
      'input-wrapper': true,
      error: this.props.hasError
    });
  }

  get arrowClass() {
    return this.expandedValue ? 'sc-icon-down-arrow' : 'sc-icon-up-arrow';
  }

  handleChange = e => this.props.onChange(e.target.value);

  render() {
    const { placeholder, lettersLeftCount, value, maxLength, hasError, disabled } = this.props;

    return (
      <div className={this.wrapperClass}>
        <div className={`${BASE_CLASS}__wrapper`}>
          <Input
            onInputRef={this.onInputRef}
            onClick={this.onInputClick}
            onChange={this.handleChange}
            value={value}
            placeholder={placeholder}
            lettersLeftCount={lettersLeftCount}
            maxLength={maxLength}
            hasError={hasError}
            disabled={disabled}
            {...getDataTestElementProps(this.props)}
          />
        </div>
        <button className={`${BASE_CLASS}__button`} onClick={this.onButtonClick}>
          <span className={this.arrowClass} />
        </button>
        {this.errors}
      </div>
    );
  }
}

AutocompleteInput.propTypes = {
  /**
   * Input value
   */
  value: PropTypes.string,

  /**
   * Is options list expanded
   */
  isExpanded: PropTypes.bool,

  /**
   * onClick callback that is called to show/hide suggestions
   */
  onClick: PropTypes.func,

  /**
   * onChange callback
   */
  onChange: PropTypes.func,

  /**
   * Input placeholder
   */
  placeholder: PropTypes.string,

  /**
   * If input is with characters limit this prop need to be passed to Input component
   */
  lettersLeftCount: PropTypes.number,

  /**
   * Defines maximum characters length
   */
  maxLength: PropTypes.number,

  /**
   * Flag if input has errors for styling
   */
  hasError: PropTypes.bool,

  /**
   * Errors list
   */
  children: PropTypes.node,

  /**
   * Callback to filter results based on input value
   */
  showFilteredOptions: PropTypes.func,

  /**
   * Callback to show all suggestions
   */
  showAllOptions: PropTypes.func,

  /**
   * Is input disabled
   */
  disabled: PropTypes.bool,

  ...dataTestElementPropTypes
};

AutocompleteInput.defaultProps = {
  value: '',
  isExpanded: false,
  onClick: () => true,
  onChange: () => true,
  placeholder: '',
  lettersLeftCount: null,
  maxLength: null,
  hasError: false,
  children: null,
  showFilteredOptions: () => true,
  showAllOptions: () => true,
  disabled: false
};
