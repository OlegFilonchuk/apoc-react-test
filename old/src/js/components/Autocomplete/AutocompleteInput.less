@import '~chameleon/less/variables.less';

@gray: #ccc;
@button-square-size: 31px;

@disabled-background: #eee;
@disabled-text: #b2b2b2;
@disabled-border: #b2b2b2;

.autocomplete-input {
  &__wrapper {
    display: inline-block;
    width: calc(100% - @button-square-size);

    .input-field__char-limit-counter {
      height: calc(100% - 8px);
      top: 4px;
      right: 4px;
      border: 1px solid @gray;
    }

    input {
      border-right: 0;
    }
  }

  &__button {
    width: @button-square-size;
    height: @button-square-size;
    border: 1px solid @primary;
    display: inline-block;
    float: right;
    text-align: center;
    cursor: pointer;

    &:focus {
      outline: none;
    }

    span {
      margin-top: 3px;
      color: @primary;
    }
  }

  &.error &__button {
    border: 1px solid @error;

    span {
      color: @error;
    }
  }

  &.disabled &__button {
    background: @disabled-background;
    border-color: @disabled-border;
    cursor: not-allowed;

    span {
      color: @disabled-text;
    }
  }
}
