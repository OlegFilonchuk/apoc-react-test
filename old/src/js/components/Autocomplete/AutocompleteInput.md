Purpose of this component is to be used together with Autocomplete.

## Standalone example

    <AutocompleteInput
      onChange={() => { console.log('Value changed') }}
      onClick={() => { console.log('Input clicked') }}
      value="some value"
      isExpanded={false}
    />
