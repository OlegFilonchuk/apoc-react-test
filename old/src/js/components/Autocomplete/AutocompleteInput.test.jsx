import React from 'react';
import { mount } from 'enzyme';

import AutocompleteInput from './AutocompleteInput';

const defaultProps = {
  onClick: () => null
};

describe('<AutocompleteInput />', () => {
  const mountAutocompleteInput = (props = {}, mountFn = mount) =>
    mountFn(<AutocompleteInput {...defaultProps} {...props} />);

  it('should render AutocompletInput component', () => {
    const autocompleteInput = mountAutocompleteInput();

    expect(autocompleteInput).toMatchSnapshot();
  });

  it('should render disabled autocomplete input component', () => {
    const disabledInput = mountAutocompleteInput({
      disabled: true
    });

    expect(disabledInput).toMatchSnapshot();
  });

  /**
   * Input actions
   */
  it('should allow to click on enabled input', () => {
    const fakeClick = jest.fn();

    const autocompleteInput = mountAutocompleteInput({
      disabled: false,
      onClick: fakeClick
    });

    autocompleteInput.find('input').simulate('click');

    expect(fakeClick).toHaveBeenCalled();
  });

  it('should not allow to click on disabled input', () => {
    const fakeClick = jest.fn();

    const autocompleteInput = mountAutocompleteInput({
      disabled: true,
      onClick: fakeClick
    });

    autocompleteInput.find('input').simulate('click');

    expect(fakeClick).not.toHaveBeenCalled();
  });

  it('should call showFilteredOptions with value on input click', () => {
    const fakeShowFilteredOptions = jest.fn();
    const fakeValue = 'fakeValue';

    const autocompleteInput = mountAutocompleteInput({
      showFilteredOptions: fakeShowFilteredOptions,
      value: fakeValue
    });

    autocompleteInput.find('input').simulate('click');

    expect(fakeShowFilteredOptions).toHaveBeenCalledWith(fakeValue);
  });

  /**
   * Button actions
   */
  it('should allow to click on enabled button', () => {
    const fakeClick = jest.fn();

    const autocompleteInput = mountAutocompleteInput({
      disabled: false,
      onClick: fakeClick
    });

    autocompleteInput.find('button').simulate('click');

    expect(fakeClick).toHaveBeenCalled();
  });

  it('should not allow to click on disabled button', () => {
    const fakeClick = jest.fn();

    const autocompleteInput = mountAutocompleteInput({
      disabled: true,
      onClick: fakeClick
    });

    autocompleteInput.find('button').simulate('click');

    expect(fakeClick).not.toHaveBeenCalled();
  });

  it('should call showAllOptions on button click', () => {
    const fakeShowAllOptions = jest.fn();

    const autocompleteInput = mountAutocompleteInput({
      showAllOptions: fakeShowAllOptions
    });

    autocompleteInput.find('button').simulate('click');

    expect(fakeShowAllOptions).toHaveBeenCalled();
  });

  /**
   * onChange
   */
  it('should call onChange callback with proper value', () => {
    const fakeOnChange = jest.fn();
    const fakeValue = 'fakeValue';
    const fakeEvent = {
      target: {
        value: fakeValue
      }
    };

    const autocompleteInput = mountAutocompleteInput({
      onChange: fakeOnChange
    });

    autocompleteInput.instance().handleChange(fakeEvent);

    expect(fakeOnChange).toHaveBeenCalledWith(fakeValue);
  });
});
