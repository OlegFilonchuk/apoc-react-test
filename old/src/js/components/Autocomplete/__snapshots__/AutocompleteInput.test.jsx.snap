// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutocompleteInput /> should render AutocompletInput component 1`] = `
<AutocompleteInput
  disabled={false}
  hasError={false}
  isExpanded={false}
  lettersLeftCount={null}
  maxLength={null}
  onChange={[Function]}
  onClick={[Function]}
  placeholder=""
  showAllOptions={[Function]}
  showFilteredOptions={[Function]}
  value=""
>
  <div
    className="autocomplete-input input-wrapper"
  >
    <div
      className="autocomplete-input__wrapper"
    >
      <Input
        className=""
        disabled={false}
        hasError={false}
        lettersLeftCount={null}
        maxLength={null}
        onChange={[Function]}
        onClick={[Function]}
        onInputRef={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        submitOnEnter={false}
        suffix=""
        type="text"
        value=""
      >
        <span
          className="input-field-wrapper"
        >
          <input
            className="sc-form-ele input-field with-characters-limit-input"
            disabled={false}
            maxLength={null}
            onChange={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="text"
            value=""
          />
          <div
            className="input-field__char-limit-counter"
          >
            <div
              className="input-field__char-limit-counter__text"
            />
          </div>
        </span>
      </Input>
    </div>
    <button
      className="autocomplete-input__button"
      onClick={[Function]}
    >
      <span
        className="sc-icon-down-arrow"
      />
    </button>
  </div>
</AutocompleteInput>
`;

exports[`<AutocompleteInput /> should render disabled autocomplete input component 1`] = `
<AutocompleteInput
  disabled={true}
  hasError={false}
  isExpanded={false}
  lettersLeftCount={null}
  maxLength={null}
  onChange={[Function]}
  onClick={[Function]}
  placeholder=""
  showAllOptions={[Function]}
  showFilteredOptions={[Function]}
  value=""
>
  <div
    className="disabled autocomplete-input input-wrapper"
  >
    <div
      className="autocomplete-input__wrapper"
    >
      <Input
        className=""
        disabled={true}
        hasError={false}
        lettersLeftCount={null}
        maxLength={null}
        onChange={[Function]}
        onClick={[Function]}
        onInputRef={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        submitOnEnter={false}
        suffix=""
        type="text"
        value=""
      >
        <span
          className="input-field-wrapper"
        >
          <input
            className="sc-form-ele input-field with-characters-limit-input"
            disabled={true}
            maxLength={null}
            onChange={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="text"
            value=""
          />
          <div
            className="input-field__char-limit-counter"
          >
            <div
              className="input-field__char-limit-counter__text"
            />
          </div>
        </span>
      </Input>
    </div>
    <button
      className="autocomplete-input__button"
      onClick={[Function]}
    >
      <span
        className="sc-icon-down-arrow"
      />
    </button>
  </div>
</AutocompleteInput>
`;
